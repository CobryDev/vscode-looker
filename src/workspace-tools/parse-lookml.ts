import * as fs from 'fs';
import * as glob from 'glob';

export interface LookmlField {
	name: String;
	type: String;
	viewName: String;
	fileName: String;
	lineNumber: Number;
}

export interface LookmlView {
	name: String;
	fields: LookmlField[];
	fileName: String;
	lineNumber: Number;
}

export interface LookmlExplore {
	name: String;
	fields: LookmlField[];
	fileName: String;
	lineNumber: Number;
}

export enum LookmlParentType {
	unknown,
	view,
	explore
}


// TODO: Add view: label parsing resulting in field names returned like "customer.id"
// rather than "id".

export class LookML {

	public views: LookmlView[] = [];
	public explores: LookmlExplore[] = [];

    public constructor() {}
    public parseWorkspaceLookmlFiles(workspacePath: String) {
		let self = this;
        return new Promise((resolve, reject) => {
            const options = {};
            glob(`${workspacePath}/**/*.view.lkml`, options, function (err: any, files: any) {
                self.findAllFieldNamesInWorkspace(files).then(result => {
					// Debugging output.
					// fs.writeFile("./view.json", JSON.stringify(self.views, undefined, 4), {mode: 0o777, flag: 'w+' }, (err) => {
					// 	if (err) {
					// 		console.error(err);
					// 		return;
					// 	}
					// 	console.log("File has been created");
					// });
                    resolve(); // TODO: Return number found to let user know if succesful.
                });
                // TODO: Should errors be handled? Maybe Unix.
              });            
        });
	}

	private async findAllFieldNamesInWorkspace(filePaths: string[]) {
		let self = this;
		return new Promise(async function(resolve, reject) {
			for (const filePath of filePaths) {
				await self.readFile(filePath);
			}
			resolve();
		});	
	}
	
	private readFile(filePath: string) {
		let self = this;
		return new Promise(async function(resolve, reject) {
			fs.readFile(filePath, 
				'utf-8',
				await function read(err, data) {
					if (err) {
						throw err;
					}
					let lines = data.split('\n');

					// TODO: Switch view name
					// TODO: Use interface objects
					// TODO: Check for explores and joins.
					// TODO: Add dimension_groups and subs

					var filename = filePath.replace(/^.*[\\\/]/, '');
					let parentType: LookmlParentType = LookmlParentType.unknown;

					let view: LookmlView = {
						'name': 'unknown',
						'fields': [],
						'fileName': '',
						'lineNumber': 0,
					};
					let explore: LookmlExplore = {
						'name': 'unknown',
						'fields': [],
						'fileName': '',
						'lineNumber': 0,
					};
					
					for (var i in lines) {
						// If commented field, skip it.
						let line = lines[i].trim();
						if (line[0] === '#' || (line[0] === '/' && line[1] === '/')) { continue ;}
						
						// Determine if save to explores or views.
						if (line.includes('view:')) {
							parentType = LookmlParentType.view;
							view = {
								'name': self.extractName(line),
								'fields': [],
								'fileName': filename,
								'lineNumber': Number(i)
							};
							continue;
						} else if (line.includes('explore:')) {
							explore = {
								'name': self.extractName(line),
								'fields': [],
								'fileName': filename,
								'lineNumber': Number(i)
							};		
							parentType = LookmlParentType.explore;
							continue;
						}
						if (line.includes('measure:') || line.includes('dimension:') || line.includes('filter:') 
						 || line.includes('parameter:') || line.includes('join:')) {
							let lookmlField: LookmlField = {
								'name': self.extractName(line),
								'type': self.extractType(line),
								'lineNumber': Number(i),
								'viewName': view.name,
								'fileName': filename
							};
							switch (parentType) {
								case LookmlParentType.explore:
									explore.fields.push(lookmlField);
									break;
								case LookmlParentType.view:
									view.fields.push(lookmlField);
									break;
								case LookmlParentType.unknown:
									console.log('CRAP! LookmlParentType is "unknown."');
									break;
								default:
									break;
							}
						}
					} // End lines loop
					if (view.name !== 'unknown') {
						self.views.push(view);
					}
					if (explore.name !== 'unknown') {
						self.explores.push(explore);
					}
					resolve();
				});
		});
	}

	private extractName(line: string) {
		var fieldName = line.substring(line.indexOf(':') + 1, line.indexOf('{') - 1).trim();
		fieldName = fieldName.replace('[^a-zA-Z0-9_]g','');
		return fieldName;
	}
	private extractType(line: string) {
		var type = line.split(':')[0];
		type = type.replace('[^a-zA-Z0-9_]g','').trim();
		return type;
	}
}

