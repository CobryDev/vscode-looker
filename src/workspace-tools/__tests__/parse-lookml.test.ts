import * as fs from "fs";
import * as path from "path";
import { LookML } from "../parse-lookml";

// Helper function to read fixture files
const readFixture = (filename: string): string => {
  return fs.readFileSync(path.join(__dirname, "fixtures", filename), "utf-8");
};

describe("LookML Parser", () => {
  describe("Simple View Parsing", () => {
    it("should correctly parse a simple view with a dimension and a measure", () => {
      // Arrange
      const lookmlContent = readFixture("01-simple-view.view.lkml");

      // Act
      const result = LookML.parseContent(
        lookmlContent,
        "01-simple-view.view.lkml"
      );

      // Assert
      expect(result.success).toBe(true);
      expect(result.data).toMatchSnapshot();
      expect(result.data?.views).toHaveLength(1);
      expect(result.data?.explores).toHaveLength(0);

      const view = result.data!.views[0];
      expect(view.name).toBe("orders");
      expect(view.fields).toHaveLength(2);

      // Verify specific fields
      const idField = view.fields.find((field) => field.name === "id");
      expect(idField?.type).toBe("dimension");

      const countField = view.fields.find((field) => field.name === "count");
      expect(countField?.type).toBe("measure");
    });
  });

  describe("Comprehensive Field Type Parsing", () => {
    it("should parse all different field types correctly", () => {
      // Arrange
      const lookmlContent = readFixture(
        "02-view-with-all-field-types.view.lkml"
      );

      // Act
      const result = LookML.parseContent(
        lookmlContent,
        "02-view-with-all-field-types.view.lkml"
      );

      // Assert
      expect(result.success).toBe(true);
      expect(result.data).toMatchSnapshot();
      expect(result.data?.views).toHaveLength(1);

      const view = result.data!.views[0];
      expect(view.name).toBe("comprehensive_test");
      expect(view.fields.length).toBeGreaterThan(5);

      // Check for different field types
      const fieldTypes = view.fields.map((field) => field.type);
      expect(fieldTypes).toContain("dimension");
      expect(fieldTypes).toContain("measure");
      expect(fieldTypes).toContain("filter");
      expect(fieldTypes).toContain("parameter");
    });
  });

  describe("Comments and Mixed Content Parsing", () => {
    it("should handle views with comments and skip commented lines", () => {
      // Arrange
      const lookmlContent = readFixture("03-view-with-comments.view.lkml");

      // Act
      const result = LookML.parseContent(
        lookmlContent,
        "03-view-with-comments.view.lkml"
      );

      // Assert
      expect(result.success).toBe(true);
      expect(result.data).toMatchSnapshot();
      expect(result.data?.views).toHaveLength(1);

      const view = result.data!.views[0];
      expect(view.name).toBe("orders_with_comments");

      // Should parse fields correctly despite comments
      const fieldNames = view.fields.map((field) => field.name);
      expect(fieldNames).toContain("id");
      expect(fieldNames).toContain("customer_id");
      expect(fieldNames).toContain("order_count");
      expect(fieldNames).toContain("total_revenue");
    });
  });

  describe("Malformed Content Handling", () => {
    it("should handle malformed LookML gracefully", () => {
      // Arrange
      const lookmlContent = readFixture("04-malformed-view.view.lkml");

      // Act
      const result = LookML.parseContent(
        lookmlContent,
        "04-malformed-view.view.lkml"
      );

      // Assert - should not throw errors, but may have incomplete parsing
      expect(result.success).toBe(true);
      expect(result.data).toMatchSnapshot();
      expect(result.data?.views).toHaveLength(1);

      const view = result.data!.views[0];
      expect(view.name).toBe("malformed_test");
      // Should still parse what it can
      expect(view.fields.length).toBeGreaterThan(0);
    });
  });

  describe("Explore Parsing", () => {
    it("should parse explores with joins correctly", () => {
      // Arrange
      const lookmlContent = readFixture("05-explore-with-joins.view.lkml");

      // Act
      const result = LookML.parseContent(
        lookmlContent,
        "05-explore-with-joins.view.lkml"
      );

      // Assert
      expect(result.success).toBe(true);
      expect(result.data).toMatchSnapshot();
      expect(result.data?.explores).toHaveLength(1);
      expect(result.data?.views).toHaveLength(0);

      const explore = result.data!.explores[0];
      expect(explore.name).toBe("sales_analysis");

      // Check for joins
      const joinFields = explore.fields.filter(
        (field) => field.type === "join"
      );
      expect(joinFields.length).toBeGreaterThan(0);

      const joinNames = joinFields.map((field) => field.name);
      expect(joinNames).toContain("customers");
      expect(joinNames).toContain("products");
    });
  });

  describe("Dimension Groups", () => {
    it("should handle dimension_group parsing", () => {
      // Arrange
      const lookmlContent = readFixture("06-dimension-groups.view.lkml");

      // Act
      const result = LookML.parseContent(
        lookmlContent,
        "06-dimension-groups.view.lkml"
      );

      // Assert
      expect(result.success).toBe(true);
      expect(result.data).toMatchSnapshot();
      expect(result.data?.views).toHaveLength(1);

      const view = result.data!.views[0];
      expect(view.name).toBe("events");
      expect(view.fields.length).toBeGreaterThan(0);

      // Verify dimension_group is parsed correctly with proper type recognition
      const dimensionGroupField = view.fields.find(
        (field) => field.type === "dimension_group"
      );
      expect(dimensionGroupField).toBeDefined();
      expect(dimensionGroupField!.name).toBe("created");
    });
  });

  describe("Edge Cases", () => {
    it("should handle empty content", () => {
      // Act
      const result = LookML.parseContent("", "empty.lkml");

      // Assert
      expect(result.success).toBe(true);
      expect(result.data?.views).toHaveLength(0);
      expect(result.data?.explores).toHaveLength(0);
    });

    it("should handle content with only comments", () => {
      // Arrange
      const commentOnlyContent = `
        # This is a comment
        # This is another comment
        # Another comment line
      `;

      // Act
      const result = LookML.parseContent(
        commentOnlyContent,
        "comments-only.lkml"
      );

      // Assert
      expect(result.success).toBe(true);
      expect(result.data?.views).toHaveLength(0);
      expect(result.data?.explores).toHaveLength(0);
    });

    it("should handle multiple views in one file", () => {
      // Arrange
      const multiViewContent = `
        view: first_view {
          dimension: id {
            type: number
          }
        }
        
        view: second_view {
          measure: count {
            type: count
          }
        }
      `;

      // Act
      const result = LookML.parseContent(multiViewContent, "multi-view.lkml");

      // Assert
      expect(result.success).toBe(true);
      expect(result.data?.views).toHaveLength(2);
      expect(result.data!.views[0].name).toBe("first_view");
      expect(result.data!.views[1].name).toBe("second_view");
    });
  });

  describe("Instance Methods", () => {
    it("should allow merging content into existing parser instance", () => {
      // Arrange
      const parser = new LookML();
      const content1 = readFixture("01-simple-view.view.lkml");
      const content2 = readFixture("02-view-with-all-field-types.view.lkml");

      // Act
      parser.parseAndMergeContent(content1, "file1.lkml");
      parser.parseAndMergeContent(content2, "file2.lkml");

      // Assert
      expect(parser.views).toHaveLength(2);
      expect(parser.views[0].name).toBe("orders");
      expect(parser.views[1].name).toBe("comprehensive_test");
    });
  });

  describe("Field Properties Validation", () => {
    it("should correctly populate field properties", () => {
      // Arrange
      const lookmlContent = readFixture("01-simple-view.view.lkml");

      // Act
      const result = LookML.parseContent(
        lookmlContent,
        "01-simple-view.view.lkml"
      );

      // Assert
      expect(result.success).toBe(true);
      const view = result.data!.views[0];
      const idField = view.fields.find((field) => field.name === "id");

      expect(idField).toBeDefined();
      expect(idField!.name).toBe("id");
      expect(idField!.type).toBe("dimension");
      expect(idField!.viewName).toBe("orders");
      expect(idField!.fileName).toBe("01-simple-view.view.lkml");
      expect(typeof idField!.lineNumber).toBe("number");
    });
  });

  describe("Error Handling", () => {
    it("should return error result for truly malformed content", () => {
      // Arrange
      const malformedContent = `
        view: test {
          dimension: id {
            type: number
            invalid_syntax_here!!!
            [unclosed bracket
          }
        }
        invalid_top_level_syntax!!!
      `;

      // Act
      const result = LookML.parseContent(malformedContent, "malformed.lkml");

      // Assert
      expect(result.success).toBe(false);
      expect(result.error).toBeDefined();
      expect(result.error?.message).toBeDefined();
      expect(result.data).toBeUndefined();
    });

    it("should propagate errors when using parseAndMergeContent", () => {
      // Arrange
      const parser = new LookML();
      const malformedContent = `
        view: test {
          dimension: id {
            invalid_syntax!!!
      `;

      // Act & Assert
      expect(() => {
        parser.parseAndMergeContent(malformedContent, "malformed.lkml");
      }).toThrow();
    });
  });
});
