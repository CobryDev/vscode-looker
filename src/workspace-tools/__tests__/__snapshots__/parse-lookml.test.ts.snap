// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LookML Parser Comments and Mixed Content Parsing should handle views with comments and skip commented lines 1`] = `
{
  "explores": [],
  "views": [
    {
      "fields": [
        {
          "fileName": "03-view-with-comments.view.lkml",
          "lineNumber": 3,
          "name": "id",
          "type": "dimension",
          "viewName": "orders_with_comments",
        },
        {
          "fileName": "03-view-with-comments.view.lkml",
          "lineNumber": 10,
          "name": "customer_id",
          "type": "dimension",
          "viewName": "orders_with_comments",
        },
        {
          "fileName": "03-view-with-comments.view.lkml",
          "lineNumber": 16,
          "name": "order_count",
          "type": "measure",
          "viewName": "orders_with_comments",
        },
        {
          "fileName": "03-view-with-comments.view.lkml",
          "lineNumber": 22,
          "name": "total_revenue",
          "type": "measure",
          "viewName": "orders_with_comments",
        },
      ],
      "fileName": "03-view-with-comments.view.lkml",
      "lineNumber": 1,
      "name": "orders_with_comments",
    },
  ],
}
`;

exports[`LookML Parser Comprehensive Field Type Parsing should parse all different field types correctly 1`] = `
{
  "explores": [],
  "views": [
    {
      "fields": [
        {
          "fileName": "02-view-with-all-field-types.view.lkml",
          "lineNumber": 1,
          "name": "id",
          "type": "dimension",
          "viewName": "comprehensive_test",
        },
        {
          "fileName": "02-view-with-all-field-types.view.lkml",
          "lineNumber": 6,
          "name": "name",
          "type": "dimension",
          "viewName": "comprehensive_test",
        },
        {
          "fileName": "02-view-with-all-field-types.view.lkml",
          "lineNumber": 11,
          "name": "status",
          "type": "dimension",
          "viewName": "comprehensive_test",
        },
        {
          "fileName": "02-view-with-all-field-types.view.lkml",
          "lineNumber": 16,
          "name": "total_count",
          "type": "measure",
          "viewName": "comprehensive_test",
        },
        {
          "fileName": "02-view-with-all-field-types.view.lkml",
          "lineNumber": 21,
          "name": "average_amount",
          "type": "measure",
          "viewName": "comprehensive_test",
        },
        {
          "fileName": "02-view-with-all-field-types.view.lkml",
          "lineNumber": 26,
          "name": "date_filter",
          "type": "filter",
          "viewName": "comprehensive_test",
        },
        {
          "fileName": "02-view-with-all-field-types.view.lkml",
          "lineNumber": 30,
          "name": "time_period",
          "type": "parameter",
          "viewName": "comprehensive_test",
        },
      ],
      "fileName": "02-view-with-all-field-types.view.lkml",
      "lineNumber": 0,
      "name": "comprehensive_test",
    },
  ],
}
`;

exports[`LookML Parser Dimension Groups should handle dimension_group parsing 1`] = `
{
  "explores": [],
  "views": [
    {
      "fields": [
        {
          "fileName": "06-dimension-groups.view.lkml",
          "lineNumber": 15,
          "name": "event_type",
          "type": "dimension",
          "viewName": "events",
        },
        {
          "fileName": "06-dimension-groups.view.lkml",
          "lineNumber": 20,
          "name": "event_count",
          "type": "measure",
          "viewName": "events",
        },
      ],
      "fileName": "06-dimension-groups.view.lkml",
      "lineNumber": 0,
      "name": "events",
    },
  ],
}
`;

exports[`LookML Parser Explore Parsing should parse explores with joins correctly 1`] = `
{
  "explores": [
    {
      "fields": [
        {
          "fileName": "05-explore-with-joins.view.lkml",
          "lineNumber": 1,
          "name": "customers",
          "type": "join",
          "viewName": "unknown",
        },
        {
          "fileName": "05-explore-with-joins.view.lkml",
          "lineNumber": 6,
          "name": "products",
          "type": "join",
          "viewName": "unknown",
        },
      ],
      "fileName": "05-explore-with-joins.view.lkml",
      "lineNumber": 0,
      "name": "sales_analysis",
    },
  ],
  "views": [],
}
`;

exports[`LookML Parser Malformed Content Handling should handle malformed LookML gracefully 1`] = `
{
  "explores": [],
  "views": [
    {
      "fields": [
        {
          "fileName": "04-malformed-view.view.lkml",
          "lineNumber": 1,
          "name": "missing_brace",
          "type": "dimension",
          "viewName": "malformed_test",
        },
        {
          "fileName": "04-malformed-view.view.lkml",
          "lineNumber": 6,
          "name": "normal_field",
          "type": "dimension",
          "viewName": "malformed_test",
        },
        {
          "fileName": "04-malformed-view.view.lkml",
          "lineNumber": 12,
          "name": "measure:",
          "type": "measure",
          "viewName": "malformed_test",
        },
      ],
      "fileName": "04-malformed-view.view.lkml",
      "lineNumber": 0,
      "name": "malformed_test",
    },
  ],
}
`;

exports[`LookML Parser Simple View Parsing should correctly parse a simple view with a dimension and a measure 1`] = `
{
  "explores": [],
  "views": [
    {
      "fields": [
        {
          "fileName": "01-simple-view.view.lkml",
          "lineNumber": 2,
          "name": "id",
          "type": "dimension",
          "viewName": "orders",
        },
        {
          "fileName": "01-simple-view.view.lkml",
          "lineNumber": 7,
          "name": "count",
          "type": "measure",
          "viewName": "orders",
        },
      ],
      "fileName": "01-simple-view.view.lkml",
      "lineNumber": 1,
      "name": "orders",
    },
  ],
}
`;
